File: Supervised Learning/Classification/Logistic Regression/coding.py
from sklearn.datasets import load_breast_cancer, load_digits
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score


def run_logistic_regression(X, y):
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )
    model = LogisticRegression(max_iter=1000)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)

    print("Accuracy:", accuracy_score(y_test, y_pred))
    print("Precision:", precision_score(y_test, y_pred, average="weighted"))
    print("Recall:", recall_score(y_test, y_pred, average="weighted"))
    print("F1 Score:", f1_score(y_test, y_pred, average="weighted"))


# Binary classification
print("Binary Classification:")
df_1 = load_breast_cancer()
X, y = df_1.data, df_1.target
run_logistic_regression(X, y)

# Multiclass classification
print("\nMulticlass Classification:")
df_2 = load_digits()
X, y = df_2.data, df_2.target
run_logistic_regression(X, y)

File: Supervised Learning/Classification/Logistic Regression/coding.R
library(caret)
library(nnet)




# Binary classification

data(BreastCancer, package = "mlbench")
set.seed(42)

X <- BreastCancer[,-1]
y <- as.numeric(BreastCancer$Class) - 1

train_index <- sample(1:nrow(BreastCancer), 0.8 * nrow(BreastCancer))
X_train <- X[train_index, ]
X_test <- X[-train_index, ]
y_train <- y[train_index]
y_test <- y[-train_index]

model <- glm(y_train ~ ., data = as.data.frame(X_train), family = "binomial")

prediction <- predict(model, newdata = as.data.frame(X_test), type = "response")
prediction <- ifelse(prediction > 0.5, 1, 0)

conf_matrix <- confusionMatrix(factor(prediction), factor(y_test))

accuracy <- conf_matrix$overall["Accuracy"]
precision <- conf_matrix$byClass["Precision"]
recall <- conf_matrix$byClass["Recall"]
f1_score <- conf_matrix$byClass["F1"]

cat("Accuracy:", accuracy, "\n")
cat("Precision:", precision, "\n")
cat("Recall:", recall, "\n")
cat("F1 Score:", f1_score, "\n")

# Multiclass classification
data(iris)

set.seed(42)

X <- iris[,-5]
y <- iris$Species


train_index <- sample(1:nrow(iris), 0.8 * nrow(iris))
X_train <- X[train_index, ]
X_test <- X[-train_index, ]
y_train <- y[train_index]
y_test <- y[-train_index]

model <- multinom(y_train ~ ., data = as.data.frame(X_train))

predictions <- predict(model, newdata = as.data.frame(X_test), type = "class")

predictions <- factor(predictions, levels = levels(y))

y_test <- factor(y_test, levels = levels(y))

conf_matrix <- confusionMatrix(predictions, y_test)

accuracy <- conf_matrix$overall["Accuracy"]
precision <- mean(conf_matrix$byClass[, "Precision"])
recall <- mean(conf_matrix$byClass[, "Recall"])
f1_score <- mean(conf_matrix$byClass[, "F1"])

cat("Accuracy:", accuracy, "\n")
cat("Precision:", precision, "\n")
cat("Recall:", recall, "\n")
cat("F1 Score:", f1_score, "\n")